Tutorial to run rippled in reporting mode (#1288)
XRPLF/xrpl-dev-portal: content/tutorials/manage-the-rippled-server/configuration/run-rippled-in-reporting-mode.md

---
html: run-rippled-in-reporting-mode.html
parent: configure-rippled.html
blurb: Run a special operating mode of rippled that handles remote procedure calls (RPC) for validated data.
labels:
  - Core Server
  - Blockchain
top_nav_grouping: Popular Pages
---
# Run rippled in Reporting Mode

> Reporting mode is a special operating mode of rippled that handles remote procedure calls (RPC) for validated data. 

Reviewer: ```suggestion
[Reporting mode](rippled-server-modes.html) is a mode of the XRP Ledger core server specialized for serving [HTTP and WebSocket APIs](rippled-api.html).
```

- Write `rippled` in monospace font except in headers. But, where appropriate, just call it the core XRP Ledger server (since it may be renamed `xrpld` or something soon anyway).
- Many people take "RPC" to mean "JSON-RPC" only, but reporting mode handles WebSocket as well.
- Instead of saying it's a "special mode"—which sounds like a weird exception—I think it's better to say it's a "specialized for" a specific purpose.

> A rippled node running in reporting mode does not connect to the P2P network, but instead extracts validated data from one or more rippled nodes that are connected to the P2P network. 

Reviewer: ```suggestion
In reporting mode, the server does not connect to the peer-to-peer network. Instead, it uses gRPC to get validated data from one or more trusted servers that are connected to the P2P network. 
```

- Avoid the word "node" where possible, since this term is particularly overloaded in the XRP Ledger (people use it to refer to objects _in_ the ledger as well as servers _running_ the ledger, and often think there's a 1:1 correlation between them, but really they're on separate layers and don't correspond to one another).

> It can then efficiently handle RPCs, reducing the load on rippled nodes running in P2P mode.

Reviewer: - A newline here doesn't actually produce a newline in the output; it'll get treated as a hard-wrapped continuation of the previous line.
- As above, put `rippled` in code font, and avoid the term "node"

> Fig: High-level illustration of how reporting mode works

Reviewer: Is there a figure somewhere? I don't see one. Also, please don't omit the . from "Fig." since otherwise non-native speakers might get confused since "fig" is a fruit, not an abbreviation.

Author: Added svg

The reporting mode of rippled uses two datastores:
The primary persistent datastore for rippled that includes transaction metadata, account states, and ledger headers. You can use NuDB or Cassandra as the primary persistent datastore.

> Postgres database to hold relational data, which is used mainly by tx and account_tx. 

Reviewer: - I suggest making the two data stores bullet points. Don't forget the blank line before the list so that it doesn't get treated as a continuation of the previous paragraph.
- "Postgres" is slang/nickname; "PostgreSQL" is the proper name.
- You can make tx and account_tx into links to the appropriate method definitions. If you include the follow code at the end of your file, it'll define links you can use throughout the page like `[tx method][]` and `[account_tx method][]`:

      <!--{# common link defs #}-->
      {% include '_snippets/rippled-api-links.md' %}			
      {% include '_snippets/tx-type-links.md' %}			
      {% include '_snippets/rippled_versions.md' %}

When a client RPC request is received by a server running in reporting mode:
The server accesses data from these datastores to respond to the request. 

> If any request requires access to the P2P network, then the reporting mode server forwards the request to a P2P node and then forwards the response from the P2P node to the client.

Reviewer: Change to active voice:

```suggestion
When a reporting mode server receives an API request, it loads the data from these data stores if possible. For requests that require data from the P2P network, the reporting mode forwards the request to a P2P server, and then passes the response back to the client.
```

> Multiple reporting nodes can share access to the same network accessible databases (Postgres and Cassandra); at any given time, only one reporting node will be performing ETL and writing to the databases, while the others simply read from the databases. 

Reviewer: - As above, use "PostgreSQL" not "Postgres".
- You can avoid jargon like "ETL" (extract, transform, and load)—the phrase "writing to the databases" already overs the important part.
- Avoid the word "simply" since it can contribute to a condescending tone. (I don't think it does here, but still a good practice to avoid it.)

```suggestion
Multiple reporting mode servers can share access to the same network accessible databases (PostgreSQL and Cassandra); at any given time, only one reporting mode server writes to the databases, while all the others read from the databases. 
```

## How to Run Reporting Mode

### Prerequisites

> 1. Ensure that your system meets the requirements to run rippled. For more information, see https://xrpl.org/system-requirements.html. 

Reviewer: ```suggestion
1. Ensure that your system meets the [system requirements](system-requirements.html).
```

>     Note: If you choose to use Cassandra as the database, the disk requirements for rippled will be lower as the data will not be stored on your local disk.  

Reviewer: ```suggestion
    **Note:** If you choose to use Cassandra as the database, the disk requirements for rippled will be lower as the data will not be stored on your local disk.  
```

Put "Note:" in bold so this becomes a [callout](https://dactyl.link/callouts.html).

> 2. In order to run reporting mode, you also need to run one or more rippled nodes in P2P mode. Ensure that you have at least one rippled node running in P2P mode. 

Reviewer: - Change "node" and `rippled` as above.
- Link the standard install instructions to set up P2P mode.

3. Install and configure the datastores required to run rippled in reporting mode. 

>     a. Install Postgres. 

Reviewer: Numbering the steps "a." "b." and so forth doesn't produce ordered lists in Markdown. Use numbers here instead.

By doing this, you may need to indent code blocks another 4 spaces since the parser will recognize those lines as being a level "deeper" than before.

        brew install postgres

>         brew services start postgres

Reviewer: `brew` is mostly used on macOS. Replace these with a package manager for supported operating system (`apt-get` on Ubuntu or `yum` on RHEL)— or maybe have tabs for both. Be sure to double-check the package names on those systems too

        psql postgres
	        CREATE ROLE newUser WITH LOGIN PASSWORD ‘password’;
            ALTER ROLE newUser CREATEDB;
        \q

>         psql postgres -U newuser

Reviewer: I don't know if PostgreSQL usernames are case-sensitive (it might depend on the filesystem—as a note, macOS and Windows' default filesystems are _not_ case sensitive but Linux's common file systems _are_) but this should probably match the case of the username from the previous lines (`newUser`)

        postgres=# create database reporting;

    b. Install and configure the database to be used as the primary persistent datastore. You can choose to use Cassandra or NuDB. 

>         *Cassandra*: Install Cassandra and then create a keyspace for rippled, with replication. While a replication factor of 3 is recommended, when running locally, replication is not needed and you can set `replication_factor` to 1.    

Reviewer: I suggest using double asterisks for bold here (single asterisks result in italics in Markdown)

        
            $ cqlsh [host] [port]
            > CREATE KEYSPACE rippled WITH REPLICATION =
            {'class' : 'SimpleStrategy', 'replication_factor' : 1    };

        *NuDB*: If you’re running rippled in reporting mode for your local network, you can choose to use NuDB instead of Cassandra as your backend database. NuDB is installed as part of your rippled build setup and does not require any additional installation steps.

>     c. On Mac OS, you need to manually install the Cassandra cpp driver. The Cassandra driver is built as part of the rippled build on all other platforms. 

Reviewer: - The official capitalization (since 2016) is "macOS".
- Use parallel structure.

```suggestion
    c. On macOS, you need to manually install the Cassandra cpp driver. On all other platforms, the Cassandra driver is built as part of the `rippled` build. 
```

        
        brew install cassandra-cpp-driver

### Procedure

1. Build rippled for reporting mode. 

> Complete instructions to build and run rippled are available on https://xrpl.org/build-run-rippled-macos.html.

Reviewer: - Either this should be part of the previous line, or they should have a blank line in between.
- You don't need to copy the step-by-step instructions for the build.
- Use actual links rather than pasted URLs. And use relative links, e.g. `[macOS Build Instructions](build-run-rippled-macos.html)` rather than full URLs: this helps the links stay within the same environment in staging/preview versions, and prevents the links from being treated as "external" links that take you off-site.
- The instructions for supported platforms like Ubuntu or RHEL should come before macOS instructions. macOS and Windows are for development only; Reporting Mode, especially, is intended for production use cases.

    a. Clone the rippled repository: 

        git clone https://github.com/ripple/rippled.git

    b. Access the repository on your local:

        cd rippled

    c. Create a build directory and access it:

        mkdir my_build
        cd my_build

    d. Build rippled for reporting mode: 

        cmake -G "Unix Makefiles" -Dreporting=ON -DCMAKE_BUILD_TYPE=Debug ..

    e. Run the build using CMake:

        cmake --build . -- -j 4

2. Create a configuration file to run rippled in reporting mode. 

>     Make a copy of the example config file,rippled.cfg, and save it as ripple-reporting-mode.cfg in a location that enables you to run rippled as a non-root user. Access the rippled directory and run:

Reviewer: - You need a blank line before this one; right now it's being treated as a continuation of the previous line.
- Missing space after the comma 
- Put file names in `monospace`
- I don't think `ripple-` is an appropriate prefix to use for the file name. How about `reporting-mode.cfg` instead?
- The sentence starting "Access..." sounds like an additional step, but what follows is actually the same as the instructions you just provided. Maybe "For example:" would be a better way of introducing the sample commands.
- You can use the default package install location (for supported OSes) in your example.

```suggestion

    Make a copy of the example config file, `rippled.cfg`, and save it as `reporting-mode.cfg` in a location that enables you to run `rippled` as a non-root user. For example:
    
        cd /etc/opt/ripple/
```

        mkdir -p $HOME/.config/ripple

        cp cfg/rippled-example.cfg $HOME/.config/ripple/rippled-reporting-mode.cfg

3. Edit rippled-reporting-mode.cfg to set necessary file paths. The user you plan to run rippled as must have write permissions to all of the paths you specify here.

>     a. Set the [node_db] path to the location where you want to store the ledger database.

Reviewer: Put config file parameters like `[node_db]` in monospace as well.

    b. Set the [database_path] to the location where you want to store other database data. (This includes an SQLite database with configuration data, and is typically one level above the [node_db] path field.)
        
    c. Set the [debug_logfile] to a path where rippled can write logging information.

Note that these are the only configurations required for rippled to start up successfully. All other configurations are optional and can be tweaked after you have a working server.

> 4. Edit rippled-reporting-mode.cfg to enable Reporting Mode by updating the [Reporting ] stanza in your rippled.cfg. 

Reviewer: - `rippled`'s config file stanza names are case-sensitive. When you mention it, you should put it in monospace and remove the trailing space too. But, here, I think it's better to give the high-level description only, since you have the specifics in sub-steps.

```suggestion
4. Edit `reporting-mode.cfg` file to enable Reporting Mode:
```

>     a. Uncomment the Reporting stanza:

Reviewer: As above, use numbers, not letters, for all levels of lists.

>     a. Uncomment the Reporting stanza:

Reviewer: ```suggestion
    a. Uncomment the `[reporting]` stanza or add a new one:
```

I suggest "or add a new one" for people who are not using the default config file as-is.

        [reporting]
        etl_source
        read_only=0

>     b. List the rippled sources (ETL sources) to extract data from:

Reviewer: Maybe clarify that these need to be `rippled` servers with gRPC enabled and you should only put servers you trust to provide accurate data here. (Since reporting mode doesn't connect to the P2P network, it can't verify that the data actually matches the network consensus ledger.)

        
        [etl_source]
        source_grpc_port=50051
        source_ws_port=6006
        source_ip=127.0.0.1

5. Configure the databases

>     a. Specify the Postgres DB for ledger_tx_tables:

Reviewer: The convention I've been following is to match the config file for whether to put brackets around stuff like `[ledger_tx_tables]` (always monospace)

        [ledger_tx_tables]
        conninfo = postgres://newuser:password@127.0.0.1/reporting
        use_tx_tables=1

    b. Specify the database for node_db.

    <!-- MULTICODE_BLOCK_START -->

    *NuDB*

        [node_db]
        type=NuDB
        path=/home/ubuntu/ripple/

        [ledger_history]
        1000000

    *Cassandra*

        [node_db]
        type=Cassandra
        path=

        [ledger_history]
        1000000

    <!-- MULTICODE_BLOCK_START -->

6. Modify the configuration for rippled to open up ports

    a. Open the public websocket port:

            [port_ws_admin_local]
            port = 6006
            ip = 127.0.0.1
            admin = 127.0.0.1
            protocol = ws

>     b. Open the gRPC port:

Reviewer: This is on the wrong server.

You don't open the gRPC port on reporting mode (as far as I know). You _definitely_ need to do this on the P2P mode server though.

		}
	]
}

2021-Dec-09 21:31:53.291502 UTC ReportingETL:NFO Starting reporting etl
2021-Dec-09 21:31:53.291605 UTC Application:NFO Application starting. Version is 1.8.1+DEBUG
2021-Dec-09 21:31:53.291747 UTC LoadManager:DBG Starting
2021-Dec-09 21:31:53.291846 UTC gRPC Server:NFO Starting gRPC server at 0.0.0.0:60051
2021-Dec-09 21:31:53.293246 UTC LedgerCleaner:DBG Started
2021-Dec-09 21:31:53.295543 UTC ReportingETL::ETLSource:DBG handleMessage : Received a message on ledger  subscription stream. Message : {
   "result" : {},
   "status" : "success",
   "type" : "response"
}
 - { validated_ledger :  , ip : 127.0.0.1 , web socket port : 6006, grpc port : 50051 }
2021-Dec-09 21:31:53.368075 UTC ReportingETL:NFO monitor : Database is empty. Will download a ledger from the network.
2021-Dec-09 21:31:53.368183 UTC ReportingETL:NFO monitor : Waiting for next ledger to be validated by network...
```

## Frequently Asked Questions

### Do I need to run more than one instance of rippled in order to use Reporting mode?

> Yes. A rippled node running in reporting mode does not connect to the P2P network, but instead extracts validated data from one or more rippled nodes that are connected to the P2P network. In addition to the rippled node running in reporting mode, you need to run at least one other rippled node that is connected to the P2P network. 

Reviewer: Mention that you need to trust the P2P mode server, since a malicious one could feed you wrong data, so that's why you have to run both yourself.

### I’ve already installed rippled. Can I update the configuration file to enable reporting mode and restart rippled? 

Currently, you need to download the source and build rippled for reporting mode. 

> There are initiatives in progress to provide packages for reporting mode.

Reviewer: You're missing the short answer, "No," here. It's good to provide this context, but it's also good to answer the question directly first.

### To run rippled in Reporting mode, I need at least one rippled node running in P2P mode in addition to rippled in Reporting mode. Does this mean I need double the disk space? 

The answer depends on the location of your primary datastore. If you use Cassandra as the primary datastore, the disk space requirement to run rippled in reporting mode will be lower as the data is not stored on your local disk. The PostgreSQL server can be remote as well.

> Lastly, the p2p mode server only needs to keep very recent history, since reporting mode will persist long term history.

Reviewer: Keep P2P capitalized for consistency.

Reviewer: Also, a suggestion for simpler phrasing:

```suggestion
Lastly, the P2P mode server only needs to keep very recent history, while the reporting mode server keeps long term history.
```

For more information on system requirements to run rippled, see https://xrpl.org/system-requirements.html. 

### How can I confirm the validity of the data that comes from the Postgres or Cassandra database?

> Rippled running in Reporting mode only serves validated data from the ETL source specified in the config file. If you are accessing someone else's rippled server in P2P mode as the ETL source, you are implicitly trusting that server. If not, you need to run your own rippled node in P2P mode.

Reviewer: Doesn't this also mean that someone could edit the data in Cassandra or PostgreSQL and reporting mode would not know that it has been tampered with? (Question for the SME)

---
html: run-rippled-in-reporting-mode.html

> parent: configure-rippled.html

Reviewer: This one is a weird case because it's part installation, part configuration. My thought is that since it doesn't build on the standard install, this tutorial should be viewed as more of a non-standard installation option, so I would put it in the "Install rippled" category rather than the "Configure rippled" category.

When a reporting mode server receives an API request, it loads the data from these data stores if possible. For requests that require data from the P2P network, the reporting mode forwards the request to a P2P server, and then passes the response back to the client.

Multiple reporting mode servers can share access to the same network accessible databases (PostgreSQL and Cassandra); at any given time, only one reporting mode server writes to the databases, while all the others read from the databases. 

## How to Run Reporting Mode

### Prerequisites

1. Ensure that your system meets the [system requirements](system-requirements.html).

    **Note:** If you choose to use Cassandra as the database, the disk requirements for `rippled` will be lower as the data will not be stored on your local disk.  

2. In order to run reporting mode, you also need to run one or more `rippled` servers in [P2P mode](install-rippled.html). Ensure that you have at least one `rippled` server running in P2P mode. 

3. Install and configure the datastores required to run `rippled` in reporting mode. 
    
    1. Install Postgres. 

        <!-- MULTICODE_BLOCK_START -->

        *Linux*

>         1. Download and install PostgreSQL on Linux. For instructions, refer to https://www.postgresqltutorial.com/install-postgresql-linux/.

Reviewer: Code tabs can only contain code, not other elements like this numbered list. You'll have to restructure this section to accommodate that.

        
        2. Connect to the PostgreSQL Database Server using psql, and create a user `newuser` and a database `reporting`.

        psql postgres
	        CREATE ROLE newuser WITH LOGIN PASSWORD ‘password’;
            ALTER ROLE newuser CREATEDB;
        \q
        psql postgres -U newuser

>         postgres=# create database reporting;

Reviewer: To get this to display as a code block inside a bulleted list, you need specific criteria:

- You need a blank line above. (You're good here.)
- To make it display as a child of the most recent list item, you need to indent 4 spaces more than the innermost list item.
- Then you need to add another 4 spaces to make it a code block. Otherwise it'll get treated as regular paragraph text.

For these lines (ignoring the multicode tabs for a moment), you're 3 numbered lists deep, which means you need to indent a total of (4*3) + 4 + 4 = 20 spaces for this code block.

I find that when you're nesting that deep, it's a sign you should divide up the document into multiple sections, each with their own header, instead.

        *Windows*

        1. Download and install PostgreSQL on Windows. For instructions, refer to https://www.postgresqltutorial.com/install-postgresql/.
        
        
        2. Launch the psql program and connect to the PostgreSQL Database Server, and create a user `newuser` and a database `reporting`. 
            psql postgres
	        CREATE ROLE newuser WITH LOGIN PASSWORD ‘password’;
            ALTER ROLE newuser CREATEDB;
        \q
        psql postgres -U newuser
        postgres=# create database reporting;

        *macOS*

        1. Download and install Postgres on macOS.  

        brew install postgres
        brew services start postgres

        2. Connect to the PostgreSQL Database Server using psql and create a user `newuser` and a database `reporting`.

        psql postgres
	        CREATE ROLE newuser WITH LOGIN PASSWORD ‘password’;
            ALTER ROLE newuser CREATEDB;
        \q
        psql postgres -U newuser
        postgres=# create database reporting;

>      <!-- MULTICODE_BLOCK_START -->

Reviewer: The end of a multicode block should be marked with `<!-- MULTICODE_BLOCK_END -->` (not `START` again)
